### PR TITLE
Fix most remaining Polonius test differences

### PR DESCRIPTION
**Pull request by matthewjasper**
This fixes most of the Polonius test differences and also avoids overflow in issue-38591.rs.

r? @nikomatsakis

### CODE REVIEW

**Comment by matthewjasper** [comment]
This shouldn't be merged until the last commit has been reverted.

File: src/librustc_mir/borrow_check/nll/type_check/liveness/trace.rs [hunk 1]
```diff
@@ -37,6 +37,7 @@ pub(super) fn trace(
     flow_inits: &mut FlowAtLocation<'tcx, MaybeInitializedPlaces<'_, 'tcx>>,
     move_data: &MoveData<'tcx>,
     live_locals: Vec<Local>,
+    drop_used: Vec<(Local, Location)>,
```
Comment (nikomatsakis) [inline src/librustc_mir/borrow_check/nll/type_check/liveness/trace.rs]: I found this pretty confusing. I finally realized that this is empty when not in polonius mode, but that was pretty non-obvious. Maybe we can rename this to like `polonius_drop_used_facts` and make it an `Option`?

**Review by nikomatsakis — CHANGES_REQUESTED** [review]
r=me with nit addressed in some way

**Pull request merged**